File: pyNeuralNetwork/neural_network/learning/validator.py
import math
import pandas as pd

from neural_network.components import Network

from .abstract_learner import AbstractLearner


class Validator(AbstractLearner):
    """Class to validate a neural network
    """

    def __init__(self, network: Network, data: pd.DataFrame, batch_size: int,
                 weighted: bool = False):
        """Constructor method

        Parameters
        ----------
        network : Network
            The neural network
        data : pd.DataFrame
            All the validation data for the `Network`
        batch_size : int
            The number of datapoints used in each epoch
        weighted : bool
            If `True` then we use the WeightedPartitioner, otherwise we use
            the standard Partitioner
        """
        super().__init__(network, data, batch_size, weighted)
        self._epoch = 0

    def validate(self, factor: int):
        """Performs validation of the network.

        Parameters
        ----------
        factor : int
            The epochs on which we need to print out the validation
        """
        self._epoch += 1
        total_loss = 0
        batch_partition = self._partitioner()
        for iteration in range(math.ceil(len(self._data)
                                         / self._batch_size)):
            batch_ids = batch_partition[iteration]
            total_loss += self.forward_pass_one_batch(batch_ids)
        loss = round(total_loss / len(self._data), 8)
        if self._epoch % factor == 0:
            print(f"Validation loss: {loss}")

        if not self._regression:
            self._update_categorical_dataframe()

        return loss

    def generate_scatter(self, title: str = ''):
        """Creates scatter plot from the data and their predicted values

        Parameters
        ----------
        title : str
            An optional title to append to the plot
        """
        super().abs_generate_scatter(phase='validation', title=title)

    def comparison_scatter(self, title: str = ''):
        """Creates scatter plot comparing predicted to actual values (for
        regressional problems only).

        Parameters
        ----------
        title : str
            An optional title to append to the plot
        """
        super().abs_comparison_scatter(phase='validation', title=title)

File: javaNeuralNetwork/src/main/java/neural_network/learning/Validator.java
package neural_network.learning;

import neural_network.components.Network;
import neural_network.util.Header;

import java.io.IOException;
import java.util.List;
import java.util.NavigableMap;

/** Class to validate a neural network.
 *
 */
public class Validator extends Learner {

    private int epoch = 0;

    /** General constructor - matches {@code Learner}.
     *
     * @param network The neural network to train.
     * @param data All the data to be passed to the {@code network}.
     * @param batchSize The number of datapoints per batch for an epoch.
     * @param weighted If {@code true} then we use the {@code WeightedPartitioner},
     *                 else we use the {@code Partitioner}.
     * @param numBins If {@code weighted} is {@code true} and the {@code network} is
     *                a regressor, then we need to specify the number of bins for the
     *                {@code WeightedPartitioner}. Otherwise, this parameter is ignored.
     */
    public Validator(Network network, NavigableMap<Header, List<String>> data, int batchSize,
                     boolean weighted, int numBins) {
        super(network, data, batchSize, weighted, numBins);
    }

    /** Default constructor for non-weighted partitions - matches {@code Learner}.
     *
     * @param network The neural network to train.
     * @param data All the data to be passed to the {@code network}.
     * @param batchSize The number of datapoints per batch for an epoch.
     */
    public Validator(Network network, NavigableMap<Header, List<String>> data, int batchSize) {
        super(network, data, batchSize);
    }

    /** This method does nothing, as we need to use the {@code validate} method instead.
     *
     */
    @Override
    public void run() {}

    /** Performs validation of the network.
     *
     * @param factor The epochs on which we need to print out the validation.
     * @return The validation loss.
     */
    double validate(int factor) {
        double totalLoss = 0.0;
        List<List<Integer>> batchPartition = getPartitioner().call();
        int itsPerEpoch = (int) Math
                .ceil((double) getNumDatapoints() / getBatchSize());
        for (int iteration = 0; iteration < itsPerEpoch; iteration ++) {
            List<Integer> batchIds = batchPartition.get(iteration);

            // Do forward pass and back propagation for this specific batch
            totalLoss += forwardPassOneBatch(batchIds);
        }
        double loss = Math.round(10000 * totalLoss / getNumDatapoints()) / 10000.0;
        if (epoch % factor == 0) {
            System.out.printf("Validation loss: %.4f%n", loss);
        }

        if (! isRegressor()) {
            updateCategoricalDataframe();
        }
        // Keep a track of the epoch and increment it here
        epoch ++;
        return loss;
    }

    /** Creates scatter plot from the data and their predicted values.
     *
     * @param title An optional title to append to the plot.
     * @throws IOException If an IO error occurs.
     */
    public void generateScatter(String title) throws IOException {
        super.generateScatter("validation", title);
    }

    /** Creates scatter plot comparing predicted to actual values (for
     * regression problems only).
     *
     * @param title An optional title to append to the plot.
     * @throws IOException If an IO error occurs.
     * @throws RuntimeException If this method is called with a categorical network,
     *                          instead user should call {@code Tester.generateConfusion()}.
     */
    public void comparisonScatter(String title) throws IOException {
        super.comparisonScatter("validation", title);
    }
}
